Add support for securityContext per deployment

nwalens opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #17744 
related: #18136 

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: #17744 
related: #18136 

This PR replaces the global "uid" and "gid" with a per deployment securityContext setting.

Fields are configured as follows:
```
podSecurity:
  default:
    securityContext:
      runAsUser: 50000
      fsGroup: 0
      runAsGroup: 0
    containerSecurityContext:
      runAsUser: 50000
      runAsGroup: 0
  webserver:
    securityContext:
      enabled: false
    containerSecurityContext:
      enabled: false
```

The default will be used in case the individual setting is disabled.

This change allows for more configurability and allows the usage of arbitrary securityContexts as permitted by the official docker images.
The issue #18136 still requires the PR #18147 for Openshift clusters since there is no option to remove securityContexts altogether.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

nwalens committed:
Add support for openshift security context constraints.

- Create a rolebinding for all required service accounts so they can use the anyuid SCC in openshift.

Signed-off-by: Ney walens De Mesquita <walens@gmail.com>

jedcunningham commented on chart/files/pod-template-file.kubernetes-helm-yaml:
```diff
@@ -45,6 +45,7 @@ spec:
   {{- end }}
   containers:
     - args: []
+      securityContext: {{- omit .Values.podSecurity.pod_template.containerSecurityContext "enabled" | default (.Values.podSecurity.default.containerSecurityContext) | toYaml | nindent 8 }}
```
Instead of this `omit` pattern, would this work instead? I think it will and is more intuitive imo.

values.yaml:
```
podSecurity.pod_template.containerSecurityContext: {}
```

and:

```suggestion
      securityContext: {{- .Values.podSecurity.pod_template.containerSecurityContext | default (.Values.podSecurity.default.containerSecurityContext) | toYaml | nindent 8 }}
```

nwalens replied:
The template was amended to reflect the change.

jedcunningham commented on chart/values.schema.json:
```diff
@@ -70,6 +70,332 @@
+                        "securityContext": {
+                            "description": "Global Pod security context as defined in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podsecuritycontext-v1-core",
+                            "type": "object",
+                            "default": "See values.yaml",
```
```suggestion
                            "default": {"runAsUser": 50000, "fsGroup": 0, "runAsGroup": 0},
```

I think we should bring the actual defaults into the docs instead.

nwalens replied:
Hi @jedcunningham, the changes were applied as requested.

jedcunningham reviewed (commented):
My initial thought is we should move the component specific params from `podSecurity.<component>.securityContext` to `<component>.podSecurity.securityContext`. Basically, don't follow the lead of `images`, but of `serviceAccount`, `command`, `args`, etc. Thoughts/input @kaxil?

jedcunningham commented on chart/files/pod-template-file.kubernetes-helm-yaml:
```diff
@@ -56,6 +56,7 @@ spec:
       image: {{ template "pod_template_image" . }}
       imagePullPolicy: {{ .Values.images.airflow.pullPolicy }}
       name: base
+      securityContext: {{- .Values.podTemplateSecurity.containerSecurityContext | default (.Values.podSecurity.containerSecurityContext) | toYaml | nindent 4 }}
```
This should use `workers.containerSecurityContext` instead of having a separate param.

jedcunningham commented on chart/files/pod-template-file.kubernetes-helm-yaml:
```diff
@@ -87,9 +88,7 @@ spec:
-  securityContext:
-    runAsUser: {{ .Values.uid }}
-    fsGroup: {{ .Values.gid }}
+  securityContext: {{- .Values.podTemplateSecurity.securityContext | default (.Values.podSecurity.securityContext) | toYaml | nindent 4 }}
```
Same here, should use `workers.securityContext`.

jedcunningham commented on chart/values.yaml:
```diff
@@ -383,6 +393,19 @@ workers:
+  # When not set, the values defined on podSecurity.securityContext will be used
+  securityContext:
+    runAsUser: 50000
+    fsGroup: 0
```
```suggestion
  securityContext: {}
  #  runAsUser: 50000
  #  fsGroup: 0
```

We shouldn't set these for all of the components, otherwise `podSecurity.securityContext` will never actually be used and isn't an easy way to set it across the board.

jedcunningham reviewed (commented):
As for `uid` and `gid`, we either need to add backwards compatibility for them, or we need to wait for `2.0.0` of the chart.

One option to add backwards compatibility might be to add a template to `_helper` to render the default, allowing the logic of using `podSecurity.securityContext` or `uid` and `gid` to be contained and implemented once. Types may make it not a simple drop-in with the `default | toYaml` you currently have, but would still be reasonable.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +610,62 @@ Create the name of the cleanup service account to use
+{{/*
+Set the default podsecurity.securityContext
+If no value is passed, defaults to .Values.uid and .Values.gid
+This function is required fr backwards compatibility
```
```suggestion
This function is required for backwards compatibility
```

jedcunningham commented on chart/templates/cleanup/cleanup-cronjob.yaml:
```diff
@@ -22,6 +22,8 @@
 {{- $affinity := or .Values.cleanup.affinity .Values.affinity }}
 {{- $tolerations := or .Values.cleanup.tolerations .Values.tolerations }}
+{{- $securityContext := or .Values.cleanup.securityContext (include "defaultSecurityContext" . | mustFromJson ) }}
+{{- $containerSecurityContext := or .Values.cleanup.containerSecurityContext (include "defaultContainerSecurityContext" . | mustFromJson ) }}
```
```suggestion
{{- $securityContext := or .Values.cleanup.securityContext (include "defaultSecurityContext" . | mustFromJson) }}
{{- $containerSecurityContext := or .Values.cleanup.containerSecurityContext (include "defaultContainerSecurityContext" . | mustFromJson) }}
```

nit

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +610,62 @@ Create the name of the cleanup service account to use
+{{- if .Values.podSecurity.containerSecurityContext -}}
+{{ .Values.podSecurity.containerSecurityContext | toJson }}
+{{- else -}}
+  {{- $result := dict "runAsUser" .Values.uid "runAsGroup" .Values.gid }}
```
I don't think we should set these at the container level. We should only set it at the pod level, otherwise choices for `podSecurity.securityContext.runAsUser` or `scheduler.podSecurity.securityContext.runAsUser` will be just be overridden by this default.

Similarly, if we want to set runAsGroup by default, we should set it only in the `defaultSecurityContext` I think.

That probably means we don't need this helper template and the logic for defaulting the container securityContext can be simpler?

Am I overlooking something?

nwalens replied:
Make sense except for gitSync I think. Should I leave gitSync with the option?

That would also mean that for gitSync we would have the RunAsUser.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +610,62 @@ Create the name of the cleanup service account to use
+  {{ .Values.podSecurity.securityContext | toJson }}
+{{- else -}}
+  {{- $result := dict "runAsUser" .Values.statsd.uid }}
+  {{- $result | toJson }}
```
Can this be consistent with how `gitsync` is handled (I think the non-dict route is more readable)?. That should also work for setting more than 1, no?

nwalens replied:
Hi @jedcunningham, another thing on the gitSync vs the rest with dict.
When a template is generated, the result is always a string which in turn the rendering of the chart does not seem to understand very well - that is the reason for the toJson and fromJason I have there.

When I pass a dict, I can then serialise and deserialise as follows:

*_helpers.yaml:*
```
{{- define "defaultSecurityContext" -}}
{{- if .Values.podSecurity.securityContext -}}
  {{ .Values.podSecurity.securityContext | toJson }}
{{- else -}}
  {{- $result := dict "runAsUser" .Values.uid "fsGroup" .Values.gid }}
  {{- $result | toJson }}
{{- end -}}
{{- end -}}
```
*webserver-deployment.yaml:*
```
{{- $securityContext := or .Values.webserver.securityContext (include "defaultSecurityContext" . | mustFromJson) }}
```

Printing strings works fine for gitSync since it is being done as a template instead of an include.
To make the rest work the same as gitSync, I would have to transfer the logic completely to _helpers, including the value we have in the values.yaml.
When I tried, it became much less readable in my opinion:

*_helpers.yaml:*
```
{{- define "defaultSecurityContext" -}}
  {{- $ := index . 0 -}}
  {{- with index . 1 }}
    {{- if .securityContext -}}
      {{ .securityContext | toYaml }}
    {{- else if $.Values.podSecurity.securityContext -}}
      {{ $.Values.podSecurity.securityContext | toYaml }}
    {{- else -}}
runAsUser: {{ $.Values.uid }}
fsGroup: {{ $.Values.gid }}
    {{- end -}}
  {{- end -}}
{{- end -}}
```
*webserver-deployment.yaml:*
```
{{- $securityContext := include "defaultSecurityContext" (list . .Values.webserver) }}
```

I had to pass arguments to the template and then evaluate the proper result and output it in a way that the rendering will be understood when building the template.

Again, I'm not sure if there is an easier solution, at least I could not find one with my limited knowledge in go templating.
Any thoughts?

nwalens replied:
Another thing that came to mind is that the worker deployment requires a **uid** and **gid** to run a chown when fixed permissions is enabled.

To do that, we need to be able to extract the keys from the yaml which was produced by the template like so:
```
- "{{ pluck "runAsUser" $containerSecurityContext | first }}:{{ pluck "runAsGroup" $containerSecurityContext | first }}"
```

I'm not sure we can do that with a plain text result from the template.

jedcunningham replied:
Fair enough, makes sense, I should have known it wasn't as simple as it seemed 👍

jedcunningham commented on chart/tests/test_security_context.py:
```diff
@@ -0,0 +1,234 @@
+                "webserver": {
+                    "defaultUser": {"enabled": True},
+                    "securityContext": {"runAsUser": 9000, "fsGroup": 90},
+                    "containerSecurityContext": {"runAsUser": 8000, "runAsGroup": 80}
```
So we don't have to define these for all of them, maybe pull these out and do something like this instead:

```
# define this above
component_contexts = {
    "securityContext": {"runAsUser": 9000, "fsGroup": 90},
    "containerSecurityContext": {"runAsUser": 8000, "runAsGroup": 80}
}
```

Then change this section (and the rest to):
```suggestion
                    **component_contexts
```

jedcunningham commented on chart/values.schema.json:
```diff
@@ -1629,6 +1666,18 @@
                     "default": {}
+                },
+                "securityContext": {
+                    "description": "Global Pod security context as defined in `Pod Security Context <https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podsecuritycontext-v1-core>`_. If not set, the values from `securityContext` will be used.",
```
```suggestion
                    "description": "Triggerer Pod security context as defined in `Pod Security Context <https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podsecuritycontext-v1-core>`_. If not set, the values from `securityContext` will be used.",
```

Or similar, and for all the others that aren't actually global?

jedcunningham commented on chart/templates/workers/worker-deployment.yaml:
```diff
@@ -111,7 +110,7 @@ spec:
             - chown
             - -R
-            - "{{ .Values.uid }}:{{ .Values.gid }}"
+            - "{{ pluck "runAsUser" $securityContext | first }}:{{ pluck "fsGroup" $securityContext | first }}"
```
Is this pluck the only reason we use the `mustFromJson | toYaml` pattern everywhere?

Might be better to simplify the common case and just handle this case slightly differently? Haven't experimented with this at all, just my off the cuff thoughts.

jedcunningham commented on chart/tests/test_security_context.py:
```diff
@@ -0,0 +1,206 @@
+        )
+
+        for index in range(len(docs)):
+            assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", docs[index])
```
```suggestion
        assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", docs[0])
```
Both here and elsewhere.

jedcunningham commented on chart/tests/test_security_context.py:
```diff
@@ -0,0 +1,206 @@
+            ],
+        )
+
+        for index in range(len(docs)):
+            assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", docs[index])
+            assert 30 == jmespath.search("spec.template.spec.securityContext.fsGroup", docs[index])
```
```suggestion
        for doc in docs:
            assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", doc)
            assert 30 == jmespath.search("spec.template.spec.securityContext.fsGroup", doc)
```

Both here and elsewhere.

dstandish commented on docs/helm-chart/production-guide.rst:
```diff
@@ -258,3 +258,140 @@ In order to enable the usage of SCCs, one must set the parameter :ref:`rbac.crea
+    template:
+      spec:
+        securityContext:     # As the securityContext was not defined in ``workers``, the values from securityContext will take priority
+          runAsUser: 5000
```
50,000?

nwalens replied:
Hey, thanks for the feedback. Those were only examples, it does make sense though that we use the default values I guess.

dstandish commented on docs/helm-chart/production-guide.rst:
```diff
@@ -258,3 +258,140 @@ In order to enable the usage of SCCs, one must set the parameter :ref:`rbac.crea
+      initContainers:
+          - name: wait-for-airflow-migrations
+            securityContext:
+              runAsUser: 1001
```
where is it getting 1001 from since it wasn't provided in config?  i guess these are not configurable?

mik-laj commented on chart/values.schema.json:
```diff
@@ -1203,6 +1216,18 @@
+                },
+                "securityContext": {
+                    "description": "Global Pod security context as defined in `Pod Security Context <https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podsecuritycontext-v1-core>`_. If not set, the values from `securityContext` will be used.",
+                    "type": "object",
```
Can you use a more strict schema?
https://github.com/apache/airflow/pull/19181

nwalens replied:
Yeah, that is great, was not aware of the possibility!
Will update it accordingly.

dstandish commented on chart/values.yaml:
```diff
@@ -384,6 +390,12 @@ workers:
       maxSurge: "100%"
       maxUnavailable: "50%"
 
+  # When not set, the values defined on podSecurity.securityContext will be used
```
re 
`podSecurity.securityContext`

it does not seem you have actually added a `podSecurity` node?

nwalens replied:
Sorry about that, I did have it in the past and since them after speaking to @jedcunningham I realised that that node was kind of useless. I will update it accordingly.

dstandish commented on docs/helm-chart/production-guide.rst:
```diff
@@ -258,3 +258,140 @@ In order to enable the usage of SCCs, one must set the parameter :ref:`rbac.crea
+          - name: worker
+            securityContext:
+              runAsUser: 1001
+              runAsGroup: 0
```
i tried this with the given config and found that runas and fs are only output for `spec.template.spec.securityContext` but not the others

nwalens replied:
That is correct. Before it was changing the containers as well but after speaking to @jedcunningham we went for changing only the global pod security instead of the securityContext for each container.

The exception is only the gitSync container as it it works as a "add-on" initContainer or sidecar in some deployments.

dstandish replied:
ok so i would just make sure that the examples make sense / consistent.  they are there to help users understand what happens given certain inputs, and if the example output is inconsistent with the behavior, or if it contains output that is not affected by or related to the particular input, then it can cause confusion.  so in this case, the container sec context doesn't seem like it should be here, but please correct me if i'm missing something.

jedcunningham commented on chart/tests/test_security_context.py:
```diff
@@ -0,0 +1,189 @@
+        )
+
+        assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", docs[0])
+        assert 30 == jmespath.search("spec.template.spec.securityContext.fsGroup", docs[0])
```
```suggestion
        for doc in docs:
            assert 3000 == jmespath.search("spec.template.spec.securityContext.runAsUser", doc)
            assert 30 == jmespath.search("spec.template.spec.securityContext.fsGroup", doc)
```

Sorry, my last set of suggestions may not have been clear. For the tests where we do have more than 1 doc (e.g. more than 1 template in `show_only` in this case), we should check them all.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +609,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for securityContext
+If no value is passed for secirityCpntext or <node>.securityContext, defaults to UID in the local node
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to UID in the local node
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +609,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for securityContext
+If no value is passed for secirityCpntext or <node>.securityContext, defaults to global uid and gid
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +609,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for workers chown for persistent storage
+If no value is passed for secirityCpntext or <node>.securityContext, defaults to global uid and gid
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -610,3 +609,50 @@ Create the name of the cleanup service account to use
+  {{- $ := index . 0 -}}
+  {{- with index . 1 }}
+    {{- if .securityContext -}}
+{{ pluck "runAsUser" .securityContext | first }}:{{ pluck "fsGroup" .securityContext | first }}
```
Does this work if only `securityContext.runAsUser` is set?

jedcunningham commented on chart/templates/workers/worker-deployment.yaml:
```diff
@@ -111,7 +110,7 @@ spec:
             - chown
             - -R
-            - "{{ .Values.uid }}:{{ .Values.gid }}"
+            - "{{ include "globalSecurityContextIds" (list . .Values.workers) }}"
```
I'd wondering if we'd be better off doing something like this instead (untested, but you get the idea):

`bash -c 'exec chown -R $(id -u):$(id -g) {{ template "airflow_logs" . }}`

Then `chown` can just use the uid/gid it's running as instead of being configured explicitly.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for securityContext
+If no value is passed for secirityContext or <node>.securityContext, defaults to global uid and gid
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for securityContext
+If no value is passed for secirityContext or <node>.securityContext, defaults to UID in the local node
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to UID in the local node
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,50 @@ Create the name of the cleanup service account to use
+
+{{/*
+Set the default value for workers chown for persistent storage
+If no value is passed for secirityContext or <node>.securityContext, defaults to global uid and gid
```
```suggestion
If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
```

ashb reviewed (commented):
I don't _think_ you have a test that covers when you have a partial global security setting (say) UID, but a local security setting only GID

Could you add that please?

dstandish commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,49 @@ Create the name of the cleanup service account to use
+Set the default value for securityContext
+If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
+*/}}
+{{- define "globalSecurityContext" -}}
```
@jedcunningham we are inconsistent with our  naming of defined templates (i.e. with respect to snake case vs camel case) do you have any thoughts about which way is the right way?

jedcunningham replied:
It ultimately doesn't matter, but as of today, the example chart Helm generates does use (namespaced) camel case.

`"example_chart.serviceAccountName"`

This is probably something worth making consistent eventually.

dstandish commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,49 @@ Create the name of the cleanup service account to use
+Set the default value for securityContext
+If no value is passed for securityContext or <node>.securityContext, defaults to global uid and gid
+*/}}
+{{- define "globalSecurityContext" -}}
```
separately, i think a better name might be `airflowSecurityContext`.   the difference between the templates `globalSecurityContext` and `localSecurityContext` is not that one is global and one is local.  why?  because the global one could actually resolve to _either_ "local" or "global" -- it will try local, else it will try global, else it will use top level uid / gid.  it seems the real difference is one uses the airflow image and the other does not.

could consider just having one defined template `securityContext` and add a parameter (to the template, i.e. using a 3rd list element) signaling whether it's the airflow image or not?  and based on that you could conditionally check the global params or not.

jedcunningham replied:
I like `airflowSecurityContext`.

I think I like having a separate `localSecurityContext`, as it does grab the uid from a different object too (`$.Values.uid` vs `.uid`).

jedcunningham replied:
(I can't think of a better name than `localSecurityContext` - I think it fits pretty well anyways)

dstandish replied:
yeah i was trying to think of one and struggled also.  `serviceSecurityContext` or just `securityContext`

nice thing is it's 100% internal and can be changed later without hurting anyone :)  (so local is fine)

but that said, we could do these: `securityContextAirflow`  and `securityContextOther`... i kindof like that

dstandish reviewed (commented):
looks good just one suggetion with the naming of the defined templates, i.e. global vs local vs airflow

nwalens committed:
Merge branch 'apache:main' into feature/securityContext

nwalens committed:
Merge branch 'apache:main' into feature/securityContext

nwalens committed:
Rename helm template

- rename globalSecurityContext to airflowSecurityContext
- rename globalSecurityContextIds to airflowSecurityContextIds
- improve comments to guide on the template usage

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,85 @@ Create the name of the cleanup service account to use
+The template can be called like so:
+   include "airflowSecurityContext" (list . .Values.webserver)
+
+Where '.' is the global varriables scope and `.Values.webserver` the local variables scope for the webserver template.
```
```suggestion
Where `.` is the global variables scope and `.Values.webserver` the local variables scope for the webserver template.
```

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -616,3 +615,85 @@ Create the name of the cleanup service account to use
+The template can be called like so:
+   include "airflowSecurityContextIds" (list . .Values.workers)
+
+Where '.' is the global varriables scope and `.Values.workers` the local variables scope for the workers template.
```
```suggestion
Where `.` is the global variables scope and `.Values.workers` the local variables scope for the workers template.
```

jedcunningham reviewed (approved):
Couple minor changes 👍

nwalens committed:
Merge branch 'apache:main' into feature/securityContext

nwalens committed:
Fix typo in comment

nwalens committed:
Fix typo

- maintain consystency by using  `  instead of  '  on comments
